src/utils/url.ts

```diff
@@ -56,3 +61,170 @@ export function addUrlQuery(url: string, query: Record<string, unknown>): string
+  }
+  /* istanbul ignore next */
+  if (typeof skylinkUrl !== "string") {
+    throw new Error(`Expected parameter skylinkUrl to be type string, was type ${typeof skylinkUrl}`);
+  }
+
+  const opts = { ...defaultDownloadOptions, ...customOptions };
+
+  const query = opts.query ?? {};
+  if (opts.download) {
+    // Set the "attachment" parameter.
+    query.attachment = true;
+  }
+  if (opts.noResponseMetadata) {
+    // Set the "no-response-metadata" parameter.
+    query["no-response-metadata"] = true;
+  }
+
+  // URL-encode the path.
+  let path = "";
+  if (opts.path) {
+    if (typeof opts.path !== "string") {
+      throw new Error(`opts.path has to be a string, ${typeof opts.path} provided`);
+    }
+
+    // Encode each element of the path separately and join them.
+    //
+    // Don't use encodeURI because it does not encode characters such as '?'
+    // etc. These are allowed as filenames on Skynet and should be encoded so
+    // they are not treated as URL separators.
+    path = opts.path
+      .split("/")
+      .map((element: string) => encodeURIComponent(element))
+      .join("/");
+  }
+
+  let url;
+  if (opts.subdomain) {
+    // Get the path from the skylink. Use the empty string if not found.
+    const skylinkPath = parseSkylink(skylinkUrl, { onlyPath: true }) ?? "";
+    // Get just the skylink.
+    let skylink = parseSkylink(skylinkUrl);
+    if (skylink === null) {
+      throw new Error(`Could not get skylink out of input '${skylinkUrl}'`);
+    }
+    // Convert the skylink (without the path) to base32.
+    skylink = convertSkylinkToBase32(skylink);
+    url = addSubdomain(portalUrl, skylink);
+    url = makeUrl(url, skylinkPath, path);
+  } else {
+    // Get the skylink including the path.
+    const skylink = parseSkylink(skylinkUrl, { includePath: true });
+    if (skylink === null) {
+      throw new Error(`Could not get skylink with path out of input '${skylinkUrl}'`);
+    }
+    // Add additional path if passed in.
+    url = makeUrl(portalUrl, opts.endpointPath, skylink, path);
+  }
+  return addUrlQuery(url, query);
+}
+
+/**
+ * Constructs the full URL for the given domain,
+ * e.g. "dac.hns" => "https://dac.hns.siasky.net"
+ *
+ * @param portalUrl - The portal URL.
+ * @param domain - Domain.
+ * @returns - The full URL for the given domain.
+ */
+export function getFullDomainUrlForPortal(portalUrl: string, domain: string): string {
+  domain = trimSuffix(domain, "/");
+  return addSubdomain(portalUrl, domain);
+}
+
+/**
+ * Extracts the domain from the given full URL,
+ * e.g. "https://dac.hns.siasky.net" => "dac.hns"
+ *
+ * @param portalUrl - The portal URL.
+ * @param fullUrl - Full URL.
+ * @returns - The extracted domain.
+ */
+export function extractDomainForPortal(portalUrl: string, fullUrl: string): string {
+  const portalUrlObj = new URL(portalUrl);
+  const fullUrlObj = new URL(fullUrl);
+
+  const domain = trimSuffix(fullUrlObj.hostname, portalUrlObj.hostname, 1);
```

Comment:
I guess it makes sense to default the `limit` to 1. I've not seen an instance yet where it's different from 1... do we need it? It looks a bit funny, from the top of my head I can't think of a scenario where you might need something like this.

Reply:
Maybe we can revisit this later since it is an internal implementation detail. I already have the tests written for it as it is and don't have the wherewithall to revisit them now. Please write an issue.

Reply:
https://github.com/SkynetHQ/skynet-js/issues/27